'''

=================================================

Milestone 3

Nama  : Zafirah Aida Adista

Batch : CODA-015-RMT

Program ini dibuat untuk melakukan eksplorasi data, data cleaning, dan data Validation
menggunakan Great Expectations. Dataset yang digunakan adalah data kepuasan penumpang 
maskapai penerbangan untuk memastikan data siap diproses ke dalam pipeline ETL.

=================================================
'''

In [1]:
# Import library pandas untuk manipulasi dan analisis data
import pandas as pd

# Import library Great Expectations untuk data validation
import great_expectations as ge

# Import FileDataContext dari Great Expectations untuk mengelola konfigurasi proyek GE
from great_expectations.data_context import FileDataContext

## Impor Data

In [5]:
file_path = "C:/Users/Dell/Downloads/train.csv"
df = pd.read_csv(file_path)

In [6]:
# Menampilkan 5 data teratas
df.head()

,id,Gender,Customer Type,Age,Type of Travel,Class,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,...,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes,Satisfaction
0,70172,Male,Loyal Customer,13,Personal Travel,Eco Plus,460,3,4,3,...,5,4,3,4,4,5,5,25,18.0,Neutral Or Dissatisfied
1,5047,Male,disloyal Customer,25,Business travel,Business,235,3,2,3,...,1,1,5,3,1,4,1,1,6.0,Neutral Or Dissatisfied
2,110028,Female,Loyal Customer,26,Business travel,Business,1142,2,2,2,...,5,4,3,4,4,4,5,0,0.0,Satisfied
3,24026,Female,Loyal Customer,25,Business travel,Business,562,2,5,5,...,2,2,5,3,1,4,2,11,9.0,Neutral Or Dissatisfied
4,119299,Male,Loyal Customer,61,Business travel,Business,214,3,3,3,...,3,3,4,4,3,3,3,0,0.0,Satisfied


## Eksplorasi Data

In [7]:
# Cek struktur data dan tipe data tiap kolom
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103904 entries, 0 to 103903
Data columns (total 24 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   id                                 103904 non-null  int64  
 1   Gender                             103904 non-null  object 
 2   Customer Type                      103904 non-null  object 
 3   Age                                103904 non-null  int64  
 4   Type of Travel                     103904 non-null  object 
 5   Class                              103904 non-null  object 
 6   Flight Distance                    103904 non-null  int64  
 7   Inflight wifi service              103904 non-null  int64  
 8   Departure/Arrival time convenient  103904 non-null  int64  
 9   Ease of Online booking             103904 non-null  int64  
 10  Gate location                      103904 non-null  int64  
 11  Food and drink                     1039

Hasil eksplorasi:
- Terdapat 24 kolom dan 103904 baris
- Terdapat 18 kolom bertipe int64, 1 kolom bertipe float64, dan 5 kolom bertipe kategorikal (object)

Kesimpulan:
Tipe data pada kolom sudah sesuai

In [8]:
# Menampilkan statistik deskriptif untuk tipe numerik
df.describe()

,id,Age,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,Gate location,Food and drink,Online boarding,Seat comfort,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes
count,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103904.000000,103594.000000
mean,64924.210502,39.379706,1189.448375,2.729683,3.060296,2.756901,2.976883,3.202129,3.250375,3.439396,3.358158,3.382363,3.351055,3.631833,3.304290,3.640428,3.286351,14.815618,15.178678
std,37463.812252,15.114964,997.147281,1.327829,1.525075,1.398929,1.277621,1.329533,1.349509,1.319088,1.332991,1.288354,1.315605,1.180903,1.265396,1.175663,1.312273,38.230901,38.698682
min,1.000000,7.000000,31.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,32533.750000,27.000000,414.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,3.000000,3.000000,3.000000,2.000000,0.000000,0.000000
50%,64856.500000,40.000000,843.000000,3.000000,3.000000,3.000000,3.000000,3.000000,3.000000,4.000000,4.000000,4.000000,4.000000,4.000000,3.000000,4.000000,3.000000,0.000000,0.000000
75%,97368.250000,51.000000,1743.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,5.000000,4.000000,4.000000,4.000000,5.000000,4.000000,5.000000,4.000000,12.000000,13.000000
max,129880.000000,85.000000,4983.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,1592.000000,1584.000000


Hasil:
- Usia: variasi rentang usia 7 – 85 tahun
- Flight Distance:Ada penerbangan jarak pendek (31 km) sampai jauh (4983 km)
- Penilaian Layanan: Rata-rata semua skor layanan ada di kisaran 3 – 3.6 menandakan pelayanan cukup baik
- Delay: Sebagian besar penerbangan tidak delay (median = 0), namun terdapat delay ekstrem (~26 jam)

In [9]:
# Mengecek Unique values pada kolom yang bertipe kategorikal

# Mengambil nama-nama kolom yang bertipe object
cat_cols = df.select_dtypes(include='object').columns

# Loop untuk setiap kolom kategorikal
for col in cat_cols:
    # Menampilkan nama kolom yang sedang dicek
    print(f"\nKolom: {col}")
    # Menampilkan semua nilai unik pada kolom tersebut
    print(df[col].unique())


Kolom: Gender
['Male' 'Female']

Kolom: Customer Type
['Loyal Customer' 'disloyal Customer']

Kolom: Type of Travel
['Personal Travel' 'Business travel']

Kolom: Class
['Eco Plus' 'Business' 'Eco']

Kolom: Satisfaction
['Neutral Or Dissatisfied' 'Satisfied']


In [10]:
# Cek missing value
df.isnull().sum()

id                                     0
Gender                                 0
Customer Type                          0
Age                                    0
Type of Travel                         0
Class                                  0
Flight Distance                        0
Inflight wifi service                  0
Departure/Arrival time convenient      0
Ease of Online booking                 0
Gate location                          0
Food and drink                         0
Online boarding                        0
Seat comfort                           0
Inflight entertainment                 0
On-board service                       0
Leg room service                       0
Baggage handling                       0
Checkin service                        0
Inflight service                       0
Cleanliness                            0
Departure Delay in Minutes             0
Arrival Delay in Minutes             310
Satisfaction                           0
dtype: int64

Terdapat 310 baris missing values di kolom "Arrival Delay in Minutes"

In [11]:
# Cek Duplicate
df.duplicated().sum()

0

Tidak ada nilai duplikat pada tiap kolom

## Transformasi

### Transformasi tipe data

In [13]:
# Mengatur kolom 'id' sebagai index DataFrame
df.set_index('id', inplace=True)

Kolom id digunakan sebagai identifier unik untuk setiap data, sehingga tidak memiliki kontribusi langsung terhadap analisis. Oleh karena itu, kolom ini tidak digunakan sebagai fitur, tetapi dijadikan sebagai index DataFrame

### Handling missing values

In [14]:
# Mengisi missing value pada kolom 'Arrival Delay' dengan nilai median
df['Arrival Delay in Minutes'].fillna(
    df['Arrival Delay in Minutes'].median(), # Menghitung nilai median sebagai pengisi missing values
    inplace=True) # Mengubah DataFrame secara langsung

Missing values tidak didrop untuk menghindari kehilangan data, dan diisi menggunakan median karena lebih tahan terhadap outlier dibandingkan mean

## Data Validation

### Inisialisasi Data Context

In [15]:
# Menginisialisasi data context di folder saat ini
context = FileDataContext.create(project_root_dir='./')

### Connect ke Datasource

In [17]:
# Mendefinisikan nama Datasource
datasource_name = 'airlines'
# Menambahkan datasource berbasis pandas ke dalam Data Context
datasource = context.sources.add_pandas(datasource_name)

# Mendefinisikan nama dataset
asset_name = 'flight-satisfaction'
# Menentukan path ke file dataset
path_to_data = 'train.csv' # Pastikan file train.csv ada di folder yang sama
# Menambahkan file CSV sebagai data asset ke dalam datasource
asset = datasource.add_csv_asset(asset_name, filepath_or_buffer=path_to_data)

# Membangun batch request dari dataset
batch_request = asset.build_batch_request()

### Membuat Expectation Suite

In [18]:
# Mendefinisikan nama expectation suite
expectation_suite_name = 'flight-expectation-suite'
# Membuat atau update expectation suite di dalam data context
context.add_or_update_expectation_suite(expectation_suite_name)

# Membuat validator untuk menghubungkan data dengan expectation suite
validator = context.get_validator(
    batch_request = batch_request,
    expectation_suite_name = expectation_suite_name)

# Menampilkan preview data melalui validator
validator.head()

Calculating Metrics:   0%|          | 0/1 [00:00<?, ?it/s]

,id,Gender,Customer Type,Age,Type of Travel,Class,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,...,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes,Satisfaction
0,70172,Male,Loyal Customer,13,Personal Travel,Eco Plus,460,3,4,3,...,5,4,3,4,4,5,5,25,18.0,Neutral Or Dissatisfied
1,5047,Male,disloyal Customer,25,Business travel,Business,235,3,2,3,...,1,1,5,3,1,4,1,1,6.0,Neutral Or Dissatisfied
2,110028,Female,Loyal Customer,26,Business travel,Business,1142,2,2,2,...,5,4,3,4,4,4,5,0,0.0,Satisfied
3,24026,Female,Loyal Customer,25,Business travel,Business,562,2,5,5,...,2,2,5,3,1,4,2,11,9.0,Neutral Or Dissatisfied
4,119299,Male,Loyal Customer,61,Business travel,Business,214,3,3,3,...,3,3,4,4,3,3,3,0,0.0,Satisfied


#### Ekspektasi 1: kolom `id` harus unique

Pada tahap ini, dilakukan pengecekan terhadap kolom `id`. Hal ini bertujuan untuk memastikan bahwa setiap baris data mewakili satu entitas penumpang yang berbeda dan tidak terdapat duplikasi pencatatan dalam sistem

In [19]:
validator.expect_column_values_to_be_unique('id')

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 103904,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": [],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

#### Ekspektasi 2: Semua rating harus antara 0 sampai 5

Pada tahap ini, dilakukan validasi terhadap seluruh kolom yang merepresentasikan penilaian kepuasan penumpang (rating). Skala penilaian yang digunakan dalam survei ini adalah **0 hingga 5**. Proses ini menggunakan perulangan (*looping*) untuk memastikan bahwa tidak ada nilai di luar rentang tersebut pada 14 kolom fasilitas maskapai.

In [20]:
# Daftar kolom yang berisi data rating tingkat kepuasan (skala 0-5)
rating_cols = [
    'Inflight wifi service',
    'Departure/Arrival time convenient',
    'Ease of Online booking',
    'Gate location',
    'Food and drink',
    'Online boarding',
    'Seat comfort',
    'Inflight entertainment',
    'On-board service',
    'Leg room service',
    'Baggage handling',
    'Checkin service',
    'Inflight service',
    'Cleanliness']

# Melakukan iterasi untuk setiap kolom dalam daftar rating_cols
for col in rating_cols:
    # Memastikan nilai pada setiap kolom rating berada di rentang min 0 dan max 5
    validator.expect_column_values_to_be_between(col, min_value=0, max_value=5)


Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

#### Ekspektasi 3: kolom `Class` hanya terdiri dari kategori "Business", "Eco", "Eco Plus"

Validasi ini diterapkan pada kolom `Class` untuk memastikan bahwa data kategori kelas penerbangan hanya berisi nilai yang telah ditentukan, yaitu **Business**, **Eco**, atau **Eco Plus**. Hal ini penting untuk menjaga konsistensi data kategorikal agar tidak ada nilai di luar standar yang masuk ke dalam sistem.

In [21]:
validator.expect_column_values_to_be_in_set('Class', ["Business", "Eco", "Eco Plus"])

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 103904,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": [],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

#### Ekspektasi 4: kolom `Flight Distance` harus bertipe integer

Validasi ini bertujuan untuk memastikan konsistensi tipe data pada kolom `Flight Distance`. Pengecekan dilakukan untuk menjamin bahwa data yang masuk bertipe numerik, sehingga tidak terjadi kesalahan saat dilakukan perhitungan matematis atau analisis statistik pada tahap berikutnya

In [22]:
validator.expect_column_values_to_be_in_type_list('Flight Distance', ['int64', 'int'])

Calculating Metrics:   0%|          | 0/1 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "observed_value": "int64"
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

#### Ekspektasi 5: kolom `Gender` harus terdiri dari 2 kategori

Validasi ini dilakukan pada kolom `Gender` untuk memastikan hanya memiliki tepat **2 nilai unik** (Male dan Female). Pengecekan ini berfungsi untuk mendeteksi nilai kosong yang tidak terdeteksi atau kesalahan input kategori baru yang tidak sesuai dengan struktur dataset.

In [23]:
validator.expect_column_unique_value_count_to_be_between(
    column='Gender', # Kolom yang akan dicek jumlah uniknya
    min_value=2,     # Batas minimum nilai unik
    max_value=2)     # Batas maksimum nilai unik

Calculating Metrics:   0%|          | 0/4 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "observed_value": 2
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

#### Ekspektasi 6: Nilai maksimum Delay harus dalam batas wajar operasional


Validasi ini dilakukan pada kolom `Departure Delay in Minutes` dan `Arrival Delay in Minutes` untuk memastikan bahwa nilai keterlambatan maksimum yang tercatat masih berada dalam batas wajar operasional penerbangan (0 hingga 2000 menit). Hal ini berfungsi untuk mendeteksi adanya anomali data atau *outlier* ekstrem yang tidak masuk akal

In [24]:
# Memastikan nilai maksimum Departure Delay masih dalam batas wajar operasional
validator.expect_column_max_to_be_between(
    'Departure Delay in Minutes',
    min_value=0,
    max_value=2000)

# Memastikan nilai maksimum Arrival Delay masih dalam batas wajar operasional
validator.expect_column_max_to_be_between(
    'Arrival Delay in Minutes',
    min_value=0,
    max_value=2000)

Calculating Metrics:   0%|          | 0/4 [00:00<?, ?it/s]

Calculating Metrics:   0%|          | 0/4 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "observed_value": 1584.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

#### Ekspektasi 7: kolom `Age` tidak boleh negatif

Pada tahap ini, dilakukan validasi terhadap kolom `Age` untuk memastikan bahwa data umur penumpang memiliki nilai yang logis. Validasi ini menetapkan batas minimum **0 tahun**, guna memastikan tidak ada nilai negatif yang masuk akibat kesalahan input data saat proses registrasi penumpang

In [25]:
# Age tidak boleh negatif
validator.expect_column_values_to_be_between('Age',min_value=0)

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 103904,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": [],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

#### Save Expectation Suite

In [26]:
validator.save_expectation_suite(discard_failed_expectations=False)

### Data Checkpoint

In [27]:
# Membuat checkpoint

checkpoint_1 = context.add_or_update_checkpoint(
    name = 'checkpoint_1',
    validator = validator,)

In [28]:
# Menjalankan checkpoint
checkpoint_result = checkpoint_1.run()

Calculating Metrics:   0%|          | 0/125 [00:00<?, ?it/s]